### PR TITLE
Fix binary search for first/last edges

### PR DESCRIPTION
**Pull request by aneeshdurg**

### CODE REVIEW

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h [hunk 1]
```diff
@@ -337,34 +337,47 @@ class LC_CSR_CSC_Labeled_Graph
...
-          } else {
+          if (adjacent_value == key) {
             if (finding_first) {
+              // We can't update the last_bounds here because mid - 1 might have
+              // value key as well.
               r = mid - 1;
             } else {
               l = mid + 1;
             }
+            continue;
```
Comment (roshandathathri) [inline libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: nit: prefer `else if` to `continue`.

Also, please clean-up `return`.  `else` is not needed after `return`

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h [hunk 2]
```diff
@@ -337,34 +337,47 @@ class LC_CSR_CSC_Labeled_Graph
...
+          // Update bounds for last only if we're searching for the first
+          // element. We know mid has value == key, so the last element has to
+          // be at or after mid. The begin bound is inclusive so we inlude mid
+          // here.
           last_begin_bound = mid;
-          last_end_bound = r + 1;
         }
 
-        if (!(opt == FindEdgeWithLabelOptions::FindAny) && mid != limit) {
+        // If we're just searching for any valid edge, then we don't need to
```
Comment (ddn0) [inline libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: can you file an issue to make a test for this?

Comment (roshandathathri) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: to make unit tests*

Comment (roshandathathri) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: Just to be clear. This entire class will be superseded by PropertGraph soon. We should unit tests for this function then.
@amberhassaan

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h [hunk 3]
```diff
@@ -337,34 +337,47 @@ class LC_CSR_CSC_Labeled_Graph
       GraphNode value =
           in_edges ? this->getInEdgeDst(mid) : this->getEdgeDst(mid);
       if (value == key) {
-        // Update bounds for last only if we're searching for the first element.
         if (finding_first) {
+          // Update bounds for last only if we're searching for the first
```
Comment (ddn0) [inline libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: the STL uses lower_bound and upper_bound to achieve the same idea. It might be more maintainable to use those (or analogues thereof) instead of merging finding first/any/last into one function.

Comment (amberhassaan) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: I agree. The complexity and corner cases are getting out of hand here. Better to rewrite it using STL's upper_bound and lower_bound.

Comment (roshandathathri) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: That's a good idea. I don't see any benefits of maintaining our own code for this. Let's use lower_bound and upper_bound directly.

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h [hunk 4]
```diff
@@ -297,119 +297,26 @@ class LC_CSR_CSC_Labeled_Graph
...
-    auto res = FindFirstOrLastEdgeWithLabel<
-        in_edges, FindEdgeWithLabelOptions::FindAny>(node, key, data);
-    if (!res) {
-      return std::nullopt;
+  class EdgeIteratorComparator {
+  public:
+    using Graph = LC_CSR_CSC_Graph<
+        NodeTy, EdgeTy, EdgeDataByValue, HasNoLockable, UseNumaAlloc,
+        HasOutOfLineLockable, FileEdgeTy>;
+    const Graph* graph_;
```
Comment (roshandathathri) [inline libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: Why not `const Graph&`?

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h [hunk 5]
```diff
@@ -422,52 +329,34 @@ class LC_CSR_CSC_Labeled_Graph
...
     // trivial check; can't be connected if degree is 0
-    if (in_edges) {
-      if (in_degrees_[node] == 0 || degrees_[key] == 0) {
-        return std::nullopt;
-      }
-    } else {
-      if (degrees_[node] == 0 || in_degrees_[key] == 0) {
-        return std::nullopt;
-      }
+    if (degrees_[node] == 0 || in_degrees_[key] == 0) {
```
Comment (roshandathathri) [inline libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: Why this change? This is wrong if `in_edges` is true.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h (hunk 4 above)
Comment (aneeshdurg) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: I just went with `*` since `this` was already a pointer type

File: libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h (hunk 5 above)
Comment (aneeshdurg) [reply libgalois/include/katana/LC_CSR_CSC_Labeled_Graph.h]: whoops! You're right, I'll add it back

**Review by ddn0 — APPROVED** [review]
the STL provides... all hail our SGI overloads

**Review by amberhassaan — APPROVED** [review]
Looks OK.
